packages/fx-core/src/core/middleware/projectMigrator.ts

```diff
@@ -0,0 +1,427 @@
+  if (!settings.activeEnvironment) {
+    settings.activeEnvironment = "dev";
+    await fs.writeFile(projectSettingPath, JSON.stringify(settings, null, 4), {
+      encoding: "UTF-8",
+    });
+  }
+}
+
+async function cleanup(projectPath: string): Promise<void> {
+  const { _, fxConfig, templateAppPackage, fxPublishProfile } = await getMultiEnvFolders(
+    projectPath
+  );
+  await fs.remove(fxConfig);
+  await fs.remove(templateAppPackage);
+  await fs.remove(fxPublishProfile);
+  await fs.remove(path.join(templateAppPackage, ".."));
+  if (checkFileExist(path.join(fxConfig, "..", "new.env.default.json"))) {
+    await fs.remove(path.join(fxConfig, "..", "new.env.default.json"));
+  }
+}
+
+async function needMigrateToArmAndMultiEnv(ctx: CoreHookContext): Promise<boolean> {
+  if (!preCheckEnvEnabled()) {
+    return false;
+  }
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  const fxExist = await fs.pathExists(path.join(inputs.projectPath as string, ".fx"));
+  if (!fxExist) {
+    return false;
+  }
+
+  const envFileExist = await checkFileExist(
+    path.join(inputs.projectPath as string, ".fx", "env.default.json")
+  );
+  const configDirExist = await fs.pathExists(
+    path.join(inputs.projectPath as string, ".fx", "configs")
+  );
+  const armParameterExist = await checkFileExist(
+    path.join(inputs.projectPath as string, ".fx", "configs", "azure.parameters.dev.json")
+  );
+  if (envFileExist && (!armParameterExist || !configDirExist)) {
+    return true;
+  }
+  return false;
+}
+
+function preCheckEnvEnabled() {
+  if (isMultiEnvEnabled() && isArmSupportEnabled() && isBicepEnvCheckerEnabled()) {
+    return true;
+  }
+  return false;
+}
+
+async function migrateArm(ctx: CoreHookContext) {
+  await removeBotConfig(ctx);
+  await generateArmTempaltesFiles(ctx);
+  await generateArmParameterJson(ctx);
+}
+
+async function removeBotConfig(ctx: CoreHookContext) {
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  const fx = path.join(inputs.projectPath as string, `.${ConfigFolderName}`);
+  const envConfig = await fs.readJson(path.join(fx, "env.default.json"));
+  if (envConfig[ResourcePlugins.Bot]) {
+    delete envConfig[ResourcePlugins.Bot];
+    envConfig[ResourcePlugins.Bot] = { wayToRegisterBot: "create-new" };
+  }
+  await fs.writeFile(path.join(fx, "new.env.default.json"), JSON.stringify(envConfig, null, 4));
+}
+
+async function generateArmTempaltesFiles(ctx: CoreHookContext) {
+  const fakeCtx: CoreHookContext = { arguments: ctx.arguments };
+  const inputs = ctx.arguments[ctx.arguments.length - 1] as Inputs;
+  const core = ctx.self as FxCore;
+
+  const fx = path.join(inputs.projectPath as string, `.${ConfigFolderName}`);
+  const fxConfig = path.join(fx, InputConfigsFolderName);
+  const templateAzure = path.join(inputs.projectPath as string, "templates", "azure");
+  await fs.ensureDir(fx);
+  await fs.ensureDir(fxConfig);
```

Comment:
can we just `await fs.ensureDir(fxConfig);` without `await fs.ensureDir(fx);` first?

Reply:
sure, thanks!

Reply:
We do not even need to ensure .fx that has been checked in `needMigrateToArmAndMultiEnv` method.